pkg/cmd/unjoin/exec.go

```diff
@@ -0,0 +1,286 @@
+// Copyright Contributors to the Open Cluster Management project
+package unjoin
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	"gopkg.in/yaml.v2"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	k8syaml "k8s.io/apimachinery/pkg/runtime/serializer/yaml"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapiv1 "k8s.io/client-go/tools/clientcmd/api/v1"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+
+	yamlutil "k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/restmapper"
+
+	"open-cluster-management.io/clusteradm/pkg/cmd/join/scenario"
+
+	"open-cluster-management.io/clusteradm/pkg/helpers"
+	"open-cluster-management.io/clusteradm/pkg/helpers/apply"
+
+	"github.com/spf13/cobra"
+)
+
+func (o *Options) complete(cmd *cobra.Command, args []string) (err error) {
+	klog.V(1).InfoS("unjoin  options:", "dry-run", o.ClusteradmFlags.DryRun, "cluster", o.clusterName, "api-server", o.hubAPIServer, o.outputFile)
+
+	o.values = Values{
+		ClusterName: o.clusterName,
+		Hub: Hub{
+			APIServer: o.hubAPIServer,
+		},
+	}
+	klog.V(3).InfoS("values:", "clusterName", o.values.ClusterName, "hubAPIServer", o.values.Hub.APIServer)
+	return nil
+}
+
+func (o *Options) validate() error {
+	if o.token == "" {
+		return fmt.Errorf("token is missing")
+	}
+	if o.values.Hub.APIServer == "" {
+		return fmt.Errorf("hub-server is misisng")
+	}
+	if o.values.ClusterName == "" {
+		return fmt.Errorf("name is missing")
+	}
+
+	return nil
+}
+
+func (o *Options) run() error {
+
+	nameSpace := "default"
+	output := make([]string, 0)
+	reader := scenario.GetScenarioResourcesReader()
+
+	//Create an unsecure bootstrap
+	bootstrapExternalConfigUnSecure := o.createExternalBootstrapConfig()
+
+	//create external client from the bootstrap
+	externalClientUnSecure, err := createExternalClientFromBootstrap(bootstrapExternalConfigUnSecure)
+	if err != nil {
+		return err
+	}
+
+	//Create the kubeconfig for the internal client
+	o.values.Hub.KubeConfig, err = o.createKubeConfig(externalClientUnSecure, bootstrapExternalConfigUnSecure)
+	if err != nil {
+		return err
+	}
+
+	// Delete the applied resource in the Managed cluster
+	fmt.Printf("Remove applied resources in the managed cluster %s ... \n", o.clusterName)
+
+	kubeClient, apiExtensionsClient, dynamicClient, err := helpers.GetClients(o.ClusteradmFlags.KubectlFactory)
+	if err != nil {
+		return err
+	}
+
+	applierBuilder := &apply.ApplierBuilder{}
+	applier := applierBuilder.WithClient(kubeClient, apiExtensionsClient, dynamicClient).Build()
+
+	files := []string{
```

Comment:
I think you only need

```
"join/klusterlets.cr.yaml",
		"join/operator.yaml",
		"join/klusterlets.crd.yaml",
		"join/namespace.yaml",
		"join/cluster_role.yaml",
		"join/cluster_role_binding.yaml",
		"join/service_account.yaml",
```

and it would be better if you delete klusterlet at first, wait until klusterlet is removed and then remove the operators.

Reply:
@qiujian16 ,Well, I have added the checking function `WaitResourceToBeDelete` to ensure  the`klusterlet` be removed, Thanks